drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1328 @@
+			   NRFX_SUCCESS) {
+			nrfx_i2s_stop();
+			interface_error_service(i2s, "Internal service error");
+			return;
+
+		}
+		if ((p_new_buffers.p_rx_buffer == NULL)
+		    && (p_new_buffers.p_tx_buffer == NULL)) {
+			nrfx_i2s_stop();
+			return;
+		}
+		i2s->buffers = p_new_buffers;
+	} else {
+		if (interface_get_state(i2s) == I2S_IF_NEEDS_RESTART) {
+			if (interface_set_state(i2s, I2S_IF_RESTARTING) != 0) {
+				interface_error_service(i2s,
+						"Internal service error");
+			}
+		} else if (interface_get_state(i2s) == I2S_IF_STOPPING) {
+			if (interface_set_state(i2s, I2S_IF_READY)) {
+				interface_error_service(i2s,
+						"Internal service error");
+			}
+		} else if (rx_str->current_state != I2S_STATE_RUNNING &&
+			 tx_str->current_state != I2S_STATE_RUNNING) {
+			if (interface_get_state(i2s) == I2S_IF_RUNNING) {
+				interface_stop(i2s);
+			}
+		}
+	}
+}
+
+/*
+ * configuration functions
+ */
+
+static void cfg_reinit(struct device *dev)
+{
+	struct i2s_nrfx_interface *i2s = get_interface(dev);
+	struct channel_str *ch_tx = &i2s->channel_tx;
+	struct channel_str *ch_rx = &i2s->channel_rx;
+
+	nrfx_i2s_stop();
+	i2s->state = I2S_IF_READY;
+	ch_tx->current_state = I2S_STATE_READY;
+	ch_rx->current_state = I2S_STATE_READY;
+}
+
+static inline nrf_i2s_mck_t cfg_get_divider(
+		struct i2s_clk_settings_t const *clk_set, u8_t word_size)
+{
+	u32_t sub_idx = (word_size >> 3) - 1;
+
+	return clk_set->divider[sub_idx];
+}
+
+static inline nrf_i2s_ratio_t cfg_get_ratio(
+		struct i2s_clk_settings_t const *clk_set, u8_t word_size)
+{
+	u32_t sub_idx = (word_size >> 3) - 1;
+
+	return clk_set->ratio[sub_idx];
+}
+
+static void cfg_match_clock_settings(nrfx_i2s_config_t *config,
+			       struct i2s_config const *i2s_cfg)
+{
+	const struct i2s_clk_settings_t i2s_clock_settings[] =
+			NRFX_I2S_AVAILABLE_CLOCK_SETTINGS;
+	u32_t des_s_r = (s32_t)i2s_cfg->frame_clk_freq;
+	u8_t nb_of_settings_array_elems = ARRAY_SIZE(i2s_clock_settings);
+	struct i2s_clk_settings_t const *chosen_settings =
+		&i2s_clock_settings[nb_of_settings_array_elems - 1];
+
+	for (u8_t i = 1; i < nb_of_settings_array_elems; ++i) {
+		if (des_s_r < i2s_clock_settings[i].frequency) {
+			u32_t diff_h =
+				i2s_clock_settings[i].frequency - des_s_r;
+			u32_t diff_l =
+				abs((s32_t)des_s_r -
+				(s32_t)i2s_clock_settings[i - 1].frequency);
+			chosen_settings = (diff_h < diff_l) ?
+					  (&i2s_clock_settings[i]) :
+					  (&i2s_clock_settings[i - 1]);
+			break;
+		}
+	}
+	config->mck_setup = cfg_get_divider(chosen_settings,
+					    i2s_cfg->word_size);
+	config->ratio = cfg_get_ratio(chosen_settings,
+				      i2s_cfg->word_size);
+}
+
+static int cfg_periph_config(struct device *dev,
+			     struct i2s_config const *i2s_cfg)
+{
+	const struct i2s_nrfx_config *const dev_const_cfg = DEV_CFG(dev);
+	struct i2s_nrfx_data *const dev_data = DEV_DATA(dev);
+	struct i2s_nrfx_interface *i2s = get_interface(dev);
+	nrfx_i2s_config_t drv_cfg;
+
+	if (i2s_cfg == NULL) {
```

Comment:
This should be rather checked by an assertion as it indicates that someone does something fundamentally wrong, not that some error condition occurred at runtime. Same applies to checking `if (dev_data == NULL)`.